kernel/src/utilities/capability_ptr.rs

```diff
@@ -66,42 +60,45 @@ impl From<usize> for CapabilityPtr {
     /// Provenance note: may have null provenance.
     #[inline]
     fn from(from: usize) -> Self {
-        Self {
-            ptr: from as *const (),
-        }
+        Self { ptr: from.into() }
+    }
+}
+
+impl From<MachineRegister> for CapabilityPtr {
+    /// Convert a register to a [`CapabilityPtr`].
+    #[inline]
+    fn from(from: MachineRegister) -> Self {
+        Self { ptr: from }
     }
 }
 
 impl UpperHex for CapabilityPtr {
-    /// Format the capability as an uppercase hex string.
-    /// Will print at least the address, and any platform specific metadata if it exists.
     #[inline]
     fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        UpperHex::fmt(&(self.ptr as usize), f)
+        UpperHex::fmt(&self.ptr, f)
     }
 }
 
 impl LowerHex for CapabilityPtr {
-    /// Format the capability as a lowercase hex string.
-    /// Will print at least the address, and any platform specific metadata if it exists.
     #[inline]
     fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
-        LowerHex::fmt(&(self.ptr as usize), f)
+        LowerHex::fmt(&self.ptr, f)
     }
 }
 
 impl AddAssign<usize> for CapabilityPtr {
     /// Increments the address of a [`CapabilityPtr`]
     #[inline]
     fn add_assign(&mut self, rhs: usize) {
-        self.ptr = (self.ptr as *const u8).wrapping_add(rhs) as *const ();
+        self.ptr =
+            ((usize::from(self.ptr) as *const u8).wrapping_add(rhs) as *const () as usize).into();
```

Comment:
This does more conversions than necessary (the `as *const()` definitely seems unnecessary -- plus I suspect that this drops provenance).

Reply:
Reduced to a single `wrapping_byte_add` in #4332 .